tools/pipesys/src/cmd/link.rs

```diff
@@ -0,0 +1,194 @@
+        let inotify = inotify_init(&self.target, WatchMask::DELETE)?;
+
+        // Create the symlink.
+        let pid = process::id();
+        let source = format!("/proc/{pid}/fd/{dir_fd}");
+        fs::symlink(&source, &self.target)
+            .await
+            .with_context(|| format!("failed to symlink {source} to {}", self.target.display()))?;
+        info!("symlinked {} to {source}", self.target.display());
+
+        inotify_wait(inotify, &self.target, &symlink_not_found).await
+    }
+}
+
+/// Choose a plausible parent directory. For an absolute path, the parent should be available. For
+/// a relative path, assume the current process directory is intended.
+fn parent_dir(path: &Path) -> Result<PathBuf> {
+    if path.is_absolute() {
+        if let Some(parent) = path.parent() {
+            return Ok(parent.to_path_buf());
+        }
+    } else if let Ok(current) = env::current_dir() {
+        return Ok(current);
+    }
+    bail!("failed to find parent directory for {}", path.display());
+}
+
+/// Helper function to create stdout and stderr streams with the same file storage.
+async fn output_streams(path: &Path) -> Result<(std::fs::File, std::fs::File)> {
+    let stdout = fs::File::create(path).await?;
+    let stderr = stdout.try_clone().await?;
+    Ok((stdout.into_std().await, stderr.into_std().await))
+}
+
+/// Returns true if a symlink exists at the path, and false otherwise.
+async fn symlink_found(path: &Path) -> bool {
+    let res = fs::symlink_metadata(path)
+        .await
+        .with_context(|| format!("failed to query metadata for {}", path.display()));
+
+    if let Ok(m) = res {
+        if m.is_symlink() {
+            trace!("found symlink for {}", path.display());
+            return true;
+        }
+    }
+
+    trace!("no symlink found for {}", path.display());
+    false
+}
+
+/// Returns false if a symlink exists at the path, and true otherwise.
+async fn symlink_not_found(path: &Path) -> bool {
+    !symlink_found(path).await
+}
+
+/// Initialize an inotify instance with the requested watch mask.
+fn inotify_init(path: &Path, watch_mask: WatchMask) -> Result<Inotify> {
+    let parent_dir = parent_dir(path)?;
+
+    let inotify = Inotify::init().with_context(|| "failed to initialize inotify")?;
+    inotify
+        .watches()
+        .add(&parent_dir, watch_mask)
+        .with_context(|| format!("failed to add watch for {}", parent_dir.display()))?;
+
+    Ok(inotify)
+}
+
+/// Wait for inotify events until the provided function returns true.
+async fn inotify_wait<'a, F>(
+    watcher: Inotify,
+    path: &'a Path,
+    check: &dyn Fn(&'a Path) -> F,
+) -> Result<()>
+where
+    F: Future<Output = bool> + 'a,
+{
+    // Check whether the provided function succeeds right away, in case the watched file was
+    // created or removed before the inotify watch could be started.
+    if check(path).await {
+        return Ok(());
+    }
```

Comment:
Do we need to perform this check after the `event_stream` is created? I'm imagining a very short race window otherwise.

Reply:
> Do we need to perform this check after the event_stream is created? I'm imagining a very short race window otherwise.

My understanding is that the kernel will publish events to the watch descriptor as soon as the `inotify.watches.add(...)` returns. On the Rust side, we won't start reading those events until we poll the event stream, but we shouldn't miss any.

Reply:
When I reviewed that function I forgot that it was called in the context of already having called `inotify_init`. Makes sense!